### PR TITLE
Modify `update-testsuite.sh` to include proposal repos

### PR DESCRIPTION
**Pull request by binji**
This only includes tests that do not match an existing spec test (or
even an older version of a spec test, since proposal repos are often not
up-to-date with the spec repo).

  spec: 992d3dda1564ff6704e1f09101918776bc33ceac
  threads: 9cd7466140f0e68a13e782a6e74d40e22ce93dcc
  nontrapping-float-to-int-conversions: 4c8b8bcc32da32ad72f1e7f6d6d8cd09d7e9d102
  multi-value: fbd70edbdb390aed09f924cf5623e774d15c8c53
  sign-extension-ops: d48dee98c10ec017209963082afad798c193c95d
  mutable-global: 323e675cf1ed27aeaeddddcfb17cda560ceff9b9

### CODE REVIEW

**Comment by binji** [comment]
Thought this might be useful, thoughts?

I'm not a very confident basher, so please look for obvious mistakes!

**Comment by sunfishcode** [comment]
Interesting! Yes, this may indeed be useful.

One thing that occurs to me that this might not handle is spec test removals and renames, where the proposals may still have the tests under the old names (and this does happen occasionally, eg. call_indirect.wast).

And if I'm reading the patch right, it also seems like if a proposal adds content to an existing test file, the new content isn't picked up. Is that right?

Brainstorm: I wonder if it would make sense to have the script attempt to merge/rebase from the spec repo first, and if that gets any conflicts, just skip the proposal and log a message about it. Then, instead of skipping files that have the same name as spec repo files, have it only skip files that have identical content to spec repo files. How crazy is that?

**Comment by binji** [comment]
> have it only skip files that have identical content to spec repo files. How crazy is that?

Crazy minds think alike, I suppose! :-) That's basically what it does now, without the initial merge/rebase step.

The current "trick" uses the SHA of the test blob and looks for it in the spec repo, so I think it should work regardless of renames, and should catch any content changes that don't exactly match an existing blob in the repo's history. I suppose if a proposal reverted a change to a file, that would not be caught. So merging from spec HEAD first could work. I'll look into it.

**Comment by binji** [comment]
OK, this seems to work pretty well. Only a few repos don't automatically merge, and it's probably better to go and fix them than to do any clever git tricks.

File: update-testsuite.sh [hunk 1]
```diff
@@ -3,22 +3,125 @@ set -e
...
-cp "$spec"/test/core/*.wast .
+try_log_and_run() {
+    echo ">>" $*
+    $*
+}
 
-for new in *.wast; do
-    git add "$new" || :
+pushdir() {
+    pushd $1 >/dev/null || exit
```
Comment (sunfishcode) [inline update-testsuite.sh]: The script uses "set -e", so it should automatically exit if pushd (or popd, below) fails.

File: update-testsuite.sh [hunk 2]
```diff
@@ -3,22 +3,125 @@ set -e
...
+        if ! git branch | grep try-merge >/dev/null; then
+            log_and_run git branch try-merge origin/master
+        fi
+        log_and_run git checkout try-merge
+
+        # Attempt to merge with spec/master.
+        log_and_run git reset origin/master --hard
+        try_log_and_run git merge -q spec/master -m "merged"
+        if [ $? -ne 0 ]; then
+            git merge --abort
```
Comment (sunfishcode) [inline update-testsuite.sh]: Does this result in a message being printed about the merge failing? I agree with what you said elsewhere that it's better to ping the proposals to update, so it'd be useful if this script makes it obvious which proposals need to be pinged :-).

Comment (binji) [reply update-testsuite.sh]: Yes, and also it collects all the failed repos and displays them at the end too.

File: update-testsuite.sh (hunk 1 above)
Comment (binji) [reply update-testsuite.sh]: That's what I thought too, but it didn't seem to be failing when I tested it out... ah OK, so [this](https://stackoverflow.com/a/4073372) seems to be why. I guess the `set -e` was getting turned off when executed in the context of an `if`... maybe? weird stuff.

Comment (sunfishcode) [reply update-testsuite.sh]: Weird indeed.

File: update-testsuite.sh (hunk 2 above)
Comment (sunfishcode) [reply update-testsuite.sh]: Cool.

**Comment by sunfishcode** [comment]
For what it's worth, I don't think it's essential that this script be written in bash; that's just what I found easiest in the first version, which was mostly just running a sequence of commands.

**Comment by binji** [comment]
Thanks, maybe I'll rewrite in python or something to make it easier to maintain. For now bash is fine. :-)

**Pull request merged**